### ISSUE TITLE
Supporting type background for RUN

### ISSUE THREAD

**Issue by Sworddragon** [systemd/systemd#4184]
### Submission type
- [ x ] Request for enhancement (RFE)
### systemd version the issue has been seen with

> 231
### Used distribution

> Ubuntu 16.10 dev

On troubleshooting why xset called from an udev rule doesn't apply specific keyboard changes anymore it turned out that the RUN key in udev rules does block until the command has finished no matter how hard it tries to run in the background. This caused the keyboard being ready only after the xset call was done and its configuration got then overriden (probably from X.org). (but this is just the user story part and with google I found several other users which encountered the same problem)

That makes me think that it could be useful if the RUN key would also support something like "background" as type to not wait until the command has finished. This could be also useful as performance optimization as commands that not pre-depend on the rule would cause the udev-processing to be faster. It seems that this could also prevent difficult situations as for example a buggy blocking call at the udev keyboard processing would prevent being able to type anything with the keyboard to correct the situation.

**Comment by grawity** [comment]
The feature request assumes running X11 tools from udev rules is supported in the first place... it is not. Such hacks are _all_ kinds of nope. If you want to set Xkb options per keyboard, set them through your DE (if it's smart enough to react to hotplug) or [Xorg.conf](https://wiki.archlinux.org/index.php/Keyboard_configuration_in_Xorg#Using_X_configuration_files) otherwise. I think there are (or used to be) udev `ENV{XKB_*}` properties for the same as well.

**Comment by Sworddragon** [comment]
The udev rule is the result of an issue with X.org (so xorg.conf would not be suitable) and using & or doing some similar tricks doesn't cause udev to continue until the process is terminated.

**Comment by grawity** [comment]
Then wouldn't it be more useful to have a fix for the X.org issue, rather than pile workarounds on top of workarounds?

**Comment by Sworddragon** [comment]
For my specific case I would prefer it but the issue is already some years old and I do not expect it to be solved within the next years. But outside from it why should udev not support running userspace processes or not having the capability to continue rule processing before the process finished?

**Comment by zdzichu** [comment]
udev is not a process supervisor. If you want background task, use proper units and SYSTEMD_WANTS=
But instead of hacks, the issue should be fixes in your X environment.

**Comment by keszybz** [comment]
Yeah, SYSTEMD_WANTS is the way to achieve such things (even though calling xset from udev is a bad idea).

**Comment by Sworddragon** [comment]
I wonder what the purpose of RUN is if it is not intended to be used by userspace processes. SYSTEMD_WANTS requires a minimum of an additional configuration file while such a minor thing could be easily inlined (and I fear a bit using SYSTEMD_WANTS instead might expose unnecessary difficulties for specific cases).

**Comment by grawity** [comment]
> I wonder what the purpose of RUN is if it is not intended to be used by userspace processes.

It is; you're just using a different definition of "userspace" than everyone else. Userspace includes _everything_ that's not part of the kernel – and there's _a lot more_ to that than just X11 GUI apps.

Specifically, the main use of RUN= is to prepare devices _before_ they get picked up by regular apps: configure interfaces, switch USB modes, load drivers, and so on. That's the reason rule processing always waits for RUN= to finish.

But when it comes to X11, neither RUN not a service nor anything else can really guarantee that your Xorg will have detected the new keyboard in time for 'xset' to run. The only 100% reliable way to do this is to _react_ to Xorg's own "new input device" events.

I recall gnome/mate-settings-daemon used to have an option to automatically run a command whenever a new device appeared; maybe it's still there. Maybe there are other similar tools.

**Comment by Sworddragon** [comment]
> I recall gnome/mate-settings-daemon used to have an option to automatically run a command whenever a new device appeared; maybe it's still there. Maybe there are other similar tools.

I have already found a workaround for the workaround. Currently the RUN part does call the shell to write a temporary .servive file and executes it with systemctl to simulate the background behavior. The difficult thing with this solution was just that udev seems to not support nested quotes of the same type (for example single quotes within single quotes that would work in a terminal do not work in an udev rule). Not sure if this is intended and at some point in the future I'm trying to enhance this with SYSTEMD_WANTS if this is here possible.
